<a href="https://colab.research.google.com/github/danielajetunmobi/flyrank/blob/main/work/notebooks/w02_ml_task_framing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-03 — Frame Your Lane as an ML Task

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane as an ML task (type)

*Classification, clustering, ranking, or scoring — which one, and why?*

**Task type: classification feeding a ranking.**

Three separate yes/no questions, each its own model:
- Will this page decline?
- Will this page recover?
- Will this page gain momentum?

They are split because each applies to a different starting group — "recover" only to pages already struggling, "momentum" only to pages that weren't. One multi-class label would blur cases that mean opposite things.

Each model outputs a probability, and that probability doubles as the sort key for the review queue: the same number answers the yes/no question and sets review order. Same shape as the reference pipeline — `scripts/03_train_model.py` trains the classifier, `04_evaluate_and_export.py` ranks on its probability.

---

> ### 📌 Amendment (2026-08-05, after ML-06) — the target becomes a single continuous value
>
> The three-target design above is **superseded**. It is replaced by one regression target:
>
> ```
> target = log( future_daily_rate / baseline_daily_rate )
> ```
>
> **Why the three binary labels go.** Four problems, all created by thresholding rather than by
> the data:
>
> 1. **The ±20% cut is arbitrary.** It was inherited from FlyRank's `trend_direction` convention,
>    not derived from anything measured here.
> 2. **Magnitude is discarded.** A page falling 21% and a page falling 95% receive the identical
>    label, so the model cannot learn that one is worse or rank it higher — which is precisely
>    what a review queue needs. Meanwhile a page at −19% and one at −21% land in *opposite*
>    classes despite being nearly identical.
> 3. **The cohort is split three ways.** Each target trains on a subset conditioned on prior
>    state; a single regression uses every row.
> 4. **It manufactured a definitional artefact.** `future_decline` excludes already-declining
>    pages by construction, which is why two buckets in ML-06 Test 1 read exactly 0.00%. That
>    confusion disappears with the threshold.
>
> Sign carries what the three labels carried: negative is decline, positive is recovery or
> momentum depending on prior state. No information is lost by collapsing them.
>
> **The log is not cosmetic.** The raw ratio runs from −100% to +94,550% — bounded below,
> unbounded above, so a handful of giants would dominate any fit. `log(ratio)` is symmetric
> about zero and treats a halving and a doubling as equal-and-opposite moves.
>
> **What this does *not* fix.** ML-06 proved the decline gradient is an artefact of dividing by
> `trend_recent_impr` — a null carrying no future information produced a **79.0-point** gradient
> against **12.6** observed. Any ratio against the recent window inherits that, including this
> one. The denominator must also change: **`baseline_daily_rate` is to be an independent
> baseline**, not the window used to select the cohort. Fixing the label *design* and fixing the
> label *denominator* are two separate corrections; this amendment is the first.
>
> **Implemented in ML-07** (`w04_baseline_score.ipynb`), where the baseline and its target are
> built together. Nothing above is rewritten — the reasoning that led here is the record.

## 2. Target or proxy

*What would you predict? Where does that label come from — observed outcome or a defined rule?*

**Grain.** One row = one (content item, decision-point date). Needs the daily fact table; the starter CSV has no daily granularity and cannot express a future window.

**Windows** (decision point 2026-03-31):
- **Features** — trailing 90 days.
- **Trend check** — the most recent 60 of those 90, split 30/30. Matches FlyRank's `trend_pct` convention (`docs/data-dictionary.md`). The sanity-check cell below tests it against a 45/45 split: they disagree on **22.3%** of pages, and this flag decides which target population a page lands in.
- **Label** — the 30 days after the decision point, compared against the recent 30-day daily rate (not the 90-day average, which hides recoveries behind stale history — tested below).

**Targets**, all observed future outcomes rather than existing rules:

| Target | Population | Positive when |
|---|---|---|
| `future_decline` | not already declining | future change ≤ −20% |
| `future_recovery` | already declining | future change ≥ +20% |
| `future_momentum` | not already declining | future change ≥ +20% |

## 3. Success metric

*One metric you can defend. What number means 'good'?*

**Primary metric: Precision@K on a per-client queue, K = 100.**

Both parameters were unknown for most of this work and are now **decided**, not inherited. They are settled from the best available public evidence and recorded here, with the reasoning, so the choice can be argued with rather than guessed at.

**Decision 1 — the queue is per client.** FlyRank's public pricing sells *"Monthly SEO audits — up to 100 pages scanned"* per account (Sparrow → 250 Owl → 500 Eagle → unlimited Phoenix), with dedicated account managers per brand. Work is organised by account, so the queue is too. Section 4 quantifies the difference on this cohort. A global queue cannot work at any realistic capacity; a per-client one can.

**Decision 2 — K = 100, cadence monthly.** 100 is the entry-tier audit size, so it is the most conservative defensible choice; larger accounts would justify 250 or 500. Monthly matches the audit cadence and lines up cleanly with the 30-day label window — no sliding-horizon problem.

**What this overturns.** Earlier drafts assumed a *global* queue at K=50 on a *weekly* cadence. K=50 is below even the smallest tier; the weekly reading came from the W4 lecture's repeated *"this week"*, which on reflection describes the team's working rhythm rather than the client deliverable. Both were wrong and both are corrected here.

**Residual uncertainty, stated plainly.** This comes from public marketing copy, not internal process documentation. "Pages scanned" in an audit may be a wider funnel than the refresh queue — content credits (500–2,000/month, *"1 credit = 1 content piece created or optimized"*) bound what actually gets *changed*, while the audit bounds what gets *looked at*. Our output is a review queue, so the audit number is the right analogue. If the operational number turns out to differ, K moves; the per-client scoping does not.

**One operational finding worth carrying forward** (quantified in section 4): a large share of clients are small enough that a single audit could scan their entire site, so ranking barely matters for them. The prioritisation problem is concentrated in the large accounts — which is also where a global queue fails worst.

**Reported alongside:** Recall@K and the base rate, over repeated splits rather than one. Per `w01`, missed declines cost more than false flags, so precision alone hides the failure that matters.

**"Good" means** beating the base rate by a clear margin on the *mean* across repeated splits — not on a single favourable seed.

## 4. The unit of analysis, as a real dataframe

*Load your lane's slice and show it: one row = one what?*

In [1]:
%pip install -q duckdb huggingface_hub pandas scikit-learn matplotlib python-dotenv

import os
import duckdb
import pandas as pd
from huggingface_hub import hf_hub_download

# Token: prefer an env var from a local, never-committed .env; otherwise prompt.
# Never hardcode a token in a cell -- this repo is public (see SETUP.md).
try:
    from dotenv import load_dotenv
    load_dotenv()
except ImportError:
    pass

token = os.environ.get("HF_TOKEN")
if not token:
    import getpass
    token = getpass.getpass("Hugging Face token (read): ")

# Only the 5 month-partitions this decision point's windows actually touch --
# iterate on a slice, never scan the full ~79M-row table for a feasibility check.
# Downloading the whole (small) partition files is more reliable here than
# streaming range-reads over hf:// -- fewer moving parts, resumable on failure.
MONTHS = ["2025-12", "2026-01", "2026-02", "2026-03", "2026-04"]
local_files = [
    hf_hub_download(
        repo_id="FlyRank/internship-warehouse",
        repo_type="dataset",
        filename=f"fact_content_daily_performance/month={m}/data_0.parquet",
        token=token,
    )
    for m in MONTHS
]

con = duckdb.connect()
file_list = ", ".join(f"'{f}'" for f in local_files)
REL = f"read_parquet([{file_list}])"

DECISION_DATE = "2026-03-31"  # fixed decision point for this feasibility check

query = f"""
WITH prior AS (
    SELECT content_hash_id,
           -- full 90-day total: NOT a model feature (dropped -- currently
           -- unused). Kept only so the rejected-baseline sanity check below
           -- can reconstruct and show real numbers for what we tested against.
           SUM(gsc_impressions) AS prior_impressions_90d_for_rejected_baseline_test,
           -- the 30/30 window: trend check AND the future-window baseline both
           -- use this cadence (matches FlyRank's own trend_pct convention).
           SUM(gsc_impressions) FILTER (
               WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 60 DAY
                 AND report_date < DATE '{DECISION_DATE}' - INTERVAL 30 DAY
           ) AS trend_baseline_impr,
           SUM(gsc_impressions) FILTER (
               WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 30 DAY
           ) AS trend_recent_impr,
           -- the wider 45/45 window, kept ONLY for the window-choice sanity
           -- check below -- never used in the actual label definitions.
           SUM(gsc_impressions) FILTER (
               WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 90 DAY
                 AND report_date < DATE '{DECISION_DATE}' - INTERVAL 45 DAY
           ) AS prior_first_half_impr,
           SUM(gsc_impressions) FILTER (
               WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 45 DAY
           ) AS prior_second_half_impr
    FROM {REL}
    WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 90 DAY
      AND report_date < DATE '{DECISION_DATE}'
    GROUP BY content_hash_id
),
future AS (
    SELECT content_hash_id, SUM(gsc_impressions) AS future_impressions
    FROM {REL}
    WHERE report_date >= DATE '{DECISION_DATE}'
      -- exclusive: `<=` spans 31 dates, not 30, and inflates the future rate
      AND report_date < DATE '{DECISION_DATE}' + INTERVAL 30 DAY
    GROUP BY content_hash_id
)
SELECT p.*, f.future_impressions
FROM prior p JOIN future f USING (content_hash_id)
-- trend_recent_impr > 0 is required for everyone, since it's the denominator
-- for both was_declining AND future_change_pct below.
WHERE p.prior_impressions_90d_for_rejected_baseline_test > 0
  AND p.trend_baseline_impr > 0 AND p.trend_recent_impr > 0
-- deterministic row order, so results reproduce across runs
ORDER BY content_hash_id
"""

df = con.sql(query).df()
print("Grain: one row = one (content item, decision-point date =", DECISION_DATE, ")")
print("Rows (content items with data in all three windows):", len(df))
df.head()


[notice] A new release of pip is available: 26.1.2 -> 26.2
[notice] To update, run: C:\Users\user\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.


C:\Users\user\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.13_qbz5n2kfra8p0\LocalCache\local-packages\Python313\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Grain: one row = one (content item, decision-point date = 2026-03-31 )
Rows (content items with data in all three windows): 134398


,content_hash_id,prior_impressions_90d_for_rejected_baseline_test,trend_baseline_impr,trend_recent_impr,prior_first_half_impr,prior_second_half_impr,future_impressions
0,content_000005d4ced12088,126.0,24.0,82.0,33.0,93.0,83.0
1,content_00007bd2985b77c3,70.0,16.0,44.0,13.0,57.0,51.0
2,content_0000cd28fbda69f3,57.0,22.0,26.0,22.0,35.0,17.0
3,content_0000d495bfbfb4a8,32.0,17.0,15.0,NaN,32.0,0.0
4,content_00014efc121d911d,130.0,13.0,108.0,22.0,108.0,97.0


**Window-choice sanity check.** Before picking the 30/30 window for `was_declining`, I tested it against a wider 45/45 split of the same 90 days -- this is that actual test, not just a claim.

In [2]:
# This sanity check gets its own local subset (needs prior_first_half_impr > 0
# to divide safely) -- it must NOT filter or otherwise change `df`, which is the
# actual population every other cell in this notebook builds on.
check = df[df["prior_first_half_impr"] > 0].copy()

check["declining_30_30"] = ((check["trend_recent_impr"] - check["trend_baseline_impr"]) / check["trend_baseline_impr"] * 100) <= -20
check["declining_45_45"] = ((check["prior_second_half_impr"] - check["prior_first_half_impr"]) / check["prior_first_half_impr"] * 100) <= -20

agreement = (check["declining_30_30"] == check["declining_45_45"]).mean()
print("Sanity-check subset (needs data in the oldest third of the 90 days too): n =", len(check))
print(f'declining rate, 30/30 (FlyRank\'s own trend_pct convention): {check["declining_30_30"].mean():.1%}')
print(f'declining rate, 45/45 (a wider split of the full 90 days):  {check["declining_45_45"].mean():.1%}')
print()
print(f"agreement between the two window choices: {agreement:.1%}")
print(f"-> the two definitions disagree on {1 - agreement:.1%} of pages -- window choice isn't",
      "cosmetic, it decides which of the three target populations a page falls",
      "into. That's why the model below uses the 30/30 window: it matches",
      "FlyRank's own established convention rather than a width we picked ourselves.")

Sanity-check subset (needs data in the oldest third of the 90 days too): n = 106930
declining rate, 30/30 (FlyRank's own trend_pct convention): 26.4%
declining rate, 45/45 (a wider split of the full 90 days):  19.2%

agreement between the two window choices: 77.7%
-> the two definitions disagree on 22.3% of pages -- window choice isn't cosmetic, it decides which of the three target populations a page falls into. That's why the model below uses the 30/30 window: it matches FlyRank's own established convention rather than a width we picked ourselves.


In [3]:
df["prior_trend_pct"] = (
    (df["trend_recent_impr"] - df["trend_baseline_impr"]) / df["trend_baseline_impr"] * 100
)
df["was_declining"] = df["prior_trend_pct"] <= -20

recent_daily = df["trend_recent_impr"] / 30
future_daily = df["future_impressions"] / 30
df["future_change_pct"] = (future_daily - recent_daily) / recent_daily * 100

df["future_decline"] = (~df["was_declining"]) & (df["future_change_pct"] <= -20)
df["future_recovery"] = (df["was_declining"]) & (df["future_change_pct"] >= 20)
df["future_momentum"] = (~df["was_declining"]) & (df["future_change_pct"] >= 20)

# Diagnostic only -- not a fourth model target.
df["stagnant"] = (df["future_change_pct"] > -20) & (df["future_change_pct"] < 20)

print("Base rates:")
print(f'  was_declining (prior state): {df["was_declining"].mean():.1%}')
print(f'  future_decline (of all pages): {df["future_decline"].mean():.1%}')
print(
    "  future_recovery (of already-declining pages, n=%d):" % df["was_declining"].sum(),
    f'{df.loc[df["was_declining"], "future_recovery"].mean():.1%}',
)
print(
    "  future_momentum (of non-declining pages, n=%d):" % (~df["was_declining"]).sum(),
    f'{df.loc[~df["was_declining"], "future_momentum"].mean():.1%}',
)
print()
print("Diagnostic -- stagnant (implicit negative class), split by prior state:")
print(
    "  stagnant, of non-declining pages:",
    f'{df.loc[~df["was_declining"], "stagnant"].mean():.1%}',
)
print(
    "  stagnant, of already-declining pages:",
    f'{df.loc[df["was_declining"], "stagnant"].mean():.1%}',
    "-- most decliners keep declining rather than sit flat",
)

Base rates:
  was_declining (prior state): 23.8%
  future_decline (of all pages): 40.8%
  future_recovery (of already-declining pages, n=31997): 26.9%
  future_momentum (of non-declining pages, n=102401): 23.4%

Diagnostic -- stagnant (implicit negative class), split by prior state:
  stagnant, of non-declining pages: 23.0%
  stagnant, of already-declining pages: 18.9% -- most decliners keep declining rather than sit flat


In [4]:
import numpy as np

# Capacity arithmetic behind the two decisions above. Uses the same cohort
# definition as w03/ML-05 so the numbers are comparable across notebooks.
per_client = con.sql(f"""
WITH p AS (SELECT content_hash_id, ANY_VALUE(client_hash_id) AS cid,
     SUM(gsc_impressions) FILTER (WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 60 DAY
                                    AND report_date <  DATE '{DECISION_DATE}' - INTERVAL 30 DAY) AS b30,
     SUM(gsc_impressions) FILTER (WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 30 DAY) AS r30
   FROM {REL}
   WHERE report_date >= DATE '{DECISION_DATE}' - INTERVAL 90 DAY AND report_date < DATE '{DECISION_DATE}'
   GROUP BY 1),
f AS (SELECT content_hash_id, SUM(gsc_impressions) AS f30 FROM {REL}
   WHERE report_date >= DATE '{DECISION_DATE}' AND report_date < DATE '{DECISION_DATE}' + INTERVAL 30 DAY
   GROUP BY 1)
SELECT p.cid, COUNT(*) AS pages,
       SUM(CASE WHEN ((f.f30/30.0) - (p.r30/30.0)) / (p.r30/30.0) <= -0.20 THEN 1 ELSE 0 END) AS declines
FROM p JOIN f USING (content_hash_id)
WHERE p.b30 > 0 AND p.r30 > 0 GROUP BY 1 ORDER BY 1
""").df()

total_declines = per_client["declines"].sum()
print(f"{len(per_client)} clients | pages per client: median {per_client['pages'].median():.0f}, "
      f"max {per_client['pages'].max():,}")
print(f"declines in one 30-day window: {total_declines:,.0f} total, "
      f"median {per_client['declines'].median():.0f} per client")
print()

print("Recall ceiling with a PERFECT model -- audit tier vs queue scope:")
print(f"{'tier':<28}{'per client':>13}{'one global queue':>19}")
for name, k in [("Sparrow  (100 pages/mo)", 100), ("Owl      (250 pages/mo)", 250),
                ("Eagle    (500 pages/mo)", 500)]:
    pc = (np.minimum(k, per_client["declines"]) /
          per_client["declines"].replace(0, np.nan)).clip(upper=1).mean()
    print(f"  {name:<26}{pc:>12.1%}{k / total_declines:>18.2%}")

fits = (per_client["pages"] <= 500).sum()
print()
print(f"Clients whose entire site fits in one 500-page audit: "
      f"{fits} of {len(per_client)} ({fits / len(per_client):.0%})")

42 clients | pages per client: median 711, max 24,418
declines in one 30-day window: 72,203 total, median 422 per client

Recall ceiling with a PERFECT model -- audit tier vs queue scope:
tier                           per client   one global queue
  Sparrow  (100 pages/mo)          44.8%             0.14%
  Owl      (250 pages/mo)          55.2%             0.35%
  Eagle    (500 pages/mo)          64.7%             0.69%

Clients whose entire site fits in one 500-page audit: 19 of 42 (45%)


**Caveat (added after ML-04).** `w03` found that **27 of 104 clients (26.0%)** have a `gsc_data_start` later than this notebook's 90-day window opens, so part of that window predates their tracking. The rates above don't account for it — some `was_declining` values may reflect a coverage gap rather than a real trend. Filter or handle these clients before relying on the exact percentages.

**Baseline-choice sanity check.** `future_change_pct` above compares the future 30 days against the most recent 30 days (the same window already used for the trend check). I first tried the full 90-day daily average as the baseline instead, and this cell is that actual test: a page's 90-day average can be inflated by an *older, no-longer-representative* period (e.g. a page that was healthy for 60 days then declined for the last 30) -- comparing recovery against that stale blended average hides real recoveries.

In [5]:
# Recompute the REJECTED 90-day-baseline version here, to compare against the
# 30-day baseline `df` already uses above (does not modify `df`).
prior_daily_90 = df["prior_impressions_90d_for_rejected_baseline_test"] / 90
future_daily = df["future_impressions"] / 30
future_change_pct_90 = (future_daily - prior_daily_90) / prior_daily_90 * 100

was_declining = df["was_declining"]
future_decline_90 = (~was_declining) & (future_change_pct_90 <= -20)
future_recovery_90 = (was_declining) & (future_change_pct_90 >= 20)
future_momentum_90 = (~was_declining) & (future_change_pct_90 >= 20)

print("future_recovery rate (of already-declining pages):")
print(f'  90-day baseline (rejected):             {future_recovery_90[was_declining].mean():.1%}')
print(f'  30-day baseline (adopted, current df):  {df.loc[was_declining, "future_recovery"].mean():.1%}')
print()
print("Agreement between the two baselines, per target:")
print(f'  future_decline:  {(df["future_decline"] == future_decline_90).mean():.1%}')
print(f'  future_recovery: {(df.loc[was_declining, "future_recovery"] == future_recovery_90[was_declining]).mean():.1%}')
print(f'  future_momentum: {(df.loc[~was_declining, "future_momentum"] == future_momentum_90[~was_declining]).mean():.1%}')

future_recovery rate (of already-declining pages):
  90-day baseline (rejected):             17.5%
  30-day baseline (adopted, current df):  26.9%

Agreement between the two baselines, per target:
  future_decline:  84.1%
  future_recovery: 89.0%
  future_momentum: 76.5%


## 5. Why ML beats a fixed rule here

*What makes the pattern too messy for an if-statement?*

**A single threshold can't separate signal from noise.**

`w01` showed `trend_pct` reaching 44,900% in the starter data — that happens when a page goes from 1 impression to a handful. "+400%" on 1→5 impressions is noise, not a trend.

A rule like "flag anything that moved ±20%" cannot distinguish:
- a high-traffic page that genuinely dropped 20% (worth a specialist's time)
- a page that went from 2 impressions to 1 (also −50%, meaningless)

Fixing that by hand means stacking conditions — volume, then age, then position — until the rule is unmaintainable. A model weighs them together and learns that the same swing means different things at different volumes.

Even *defining* the starting point is unstable: two reasonable trend windows (30/30 vs 45/45) disagree on **22.3%** of pages.

**Scale of each population:** 23.8% of pages were already declining at this decision point; of the rest, 23.4% went on to accelerate; and 26.9% of decliners recovered — three common, distinct groups that one fixed condition can't separate.

## Self-check

Before you submit, confirm each line honestly:

- [x] Every section above is filled — markdown thinking AND the code that backs it
- [x] The notebook runs top to bottom with no errors (Runtime → Run all)
- [x] No client names, URLs, or private queries anywhere
- [x] My claims use careful words: observed, measured, directional, decision-support
- [x] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.

